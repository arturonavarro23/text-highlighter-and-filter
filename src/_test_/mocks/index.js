
export const defaultState = {
  text: 'Test words',
  colors: ['red', 'yellow', 'green'],
  selectedColor: 'red',
  selectedFilters: ['red', 'yellow'],
  ranges: []
};
