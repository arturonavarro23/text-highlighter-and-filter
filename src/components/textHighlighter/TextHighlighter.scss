@import '../../scss/abstracts/variables';

.text-highlighter {
  margin: 20px 0;
  position: relative;

  .text-editor {
    font-family: $font-family;
    font-size: 15px;
    height: 250px;
    width: 100%;
    z-index: -1;
    overflow: scroll;
    white-space: pre-wrap;
    background: transparent;
  }

  .display {
    position: absolute;
    color: transparent;
    width: 100%;
    padding: 4px;
    transform: translateY(calc(-100% - 4px));
    flex-direction: column;

    .green {
      background-color: $green-color;
      color: inherit;
    }

    .yellow {
      background-color: $yellow-color;
      color: inherit;
    }

    .red {
      background-color: $red-color;
      color: inherit;
    }
  }
}
