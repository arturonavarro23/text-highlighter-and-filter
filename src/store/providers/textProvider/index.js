import React, { createContext, useContext, useReducer } from 'react';

export const TextContext = createContext();

export const TextProvider = ({ reducer, initialState, children }) => (
  <TextContext.Provider value={useReducer(reducer, initialState)}>
    {children}
  </TextContext.Provider>
);

export const useStateValue = () => useContext(TextContext);
