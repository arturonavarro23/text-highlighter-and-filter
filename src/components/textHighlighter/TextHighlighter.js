import React, { useRef } from 'react';
import { useStateValue } from '../../store/providers/textProvider';
import { setText, modifyRanges } from '../../store/actions';
import { getRanges, highlightText } from '../../utils';

import './TextHighlighter.scss';

const TextHighlighter = () => {
  const [{ selectedColor, ranges, text }, dispatch] = useStateValue();
  const editorRef = useRef(null);
  const textAreaRef = useRef(null);

  const onScroll = e => {
    editorRef.current.scrollTop = e.target.scrollTop;
  };

  const onChange = e => {
    dispatch(setText(e.target.value));
  };

  const onMouseUp = () => {
    const { selectionStart, selectionEnd } = textAreaRef.current;
    // Check if user just click the textarea
    if (selectionStart !== selectionEnd && selectedColor) {
      setRange(selectionStart, selectionEnd);
    }
  };

  const setRange = (start, end) => {
    const newRanges = getRanges({ start, end, color: selectedColor }, ranges);
    dispatch(modifyRanges(newRanges));
  };

  return (
    <div className="text-highlighter">
      <textarea
        className="text-editor"
        ref={textAreaRef}
        value={text}
        onChange={onChange}
        onMouseUp={onMouseUp}
        onScroll={onScroll}
      />
      <div
        ref={editorRef}
        className="text-editor display"
        dangerouslySetInnerHTML={{
          __html: highlightText(text, ranges),
        }}
      />
    </div>
  );
}

export default TextHighlighter;
