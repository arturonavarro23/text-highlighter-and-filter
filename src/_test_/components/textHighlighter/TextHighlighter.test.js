import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import * as TextProvider from '../../../store/providers/textProvider';
import TextHighlighter from '../../../components/textHighlighter';
import { defaultState } from '../../mocks';

describe('<TextHighlighter> test', () => {
  it('Snapshot should render correctly', () => {
    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [defaultState]);

    const textHighlighterTree = renderer
      .create(<TextHighlighter />)
      .toJSON();

    expect(textHighlighterTree).toMatchSnapshot();
  });
  it('Text area should display expexted text', () => {
    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [defaultState]);

    const textHighlighter = mount(<TextHighlighter />);
    const textArea = textHighlighter.find("textarea");
    expect(textArea).toHaveLength(1);
    expect(textArea.text()).toBe('Test words');
  });
  it('Should display highlights', () => {

    const newState = Object.assign({}, defaultState, {
      ranges: [{ start: 0, end: 10, color: 'yellow'}],
    });

    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [newState]);

    const textHighlighter = mount(<TextHighlighter />);
    const display = textHighlighter.find('.display');
    expect(display).toHaveLength(1);
    expect(display.html()).toBe('<div class="text-editor display"><span class="yellow">Test words</span></div>');
  });
});
