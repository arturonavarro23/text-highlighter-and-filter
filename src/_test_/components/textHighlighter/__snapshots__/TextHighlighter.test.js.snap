// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextHighlighter> test Snapshot should render correctly 1`] = `
<div
  className="text-highlighter"
>
  <textarea
    className="text-editor"
    onChange={[Function]}
    onMouseUp={[Function]}
    onScroll={[Function]}
    value="Test words"
  />
  <div
    className="text-editor display"
    dangerouslySetInnerHTML={
      Object {
        "__html": undefined,
      }
    }
  />
</div>
`;
